### PR TITLE
[TorchToLinalg] Add first-class PT2E per-channel quantization support

### PR DESCRIPTION
**Pull request by ziereis**
Implements the TorchToLinalg portion of the per-channel quantization milestone described in #4600.

This adds first-class Torch dialect operations for:

- `quantized_decomposed.quantize_per_channel`
- `quantized_decomposed.dequantize_per_channel`

The TorchToLinalg conversion lowers both operations to `linalg.generic`. Per-channel scales and zero points use an axis-aware indexing map. Negative axes are normalized before constructing the maps.

The scalar quantize and dequantize payloads are shared with the existing per-tensor lowerings. The lowering supports:

- signed and unsigned quantized tensors
- scale dtype extension and truncation
- arbitrary channel axes
- optional zero points for symmetric per-channel dequantization

When `zero_points` is `None`, the dequantization generic omits both the zero-point operand/indexing map and the subtraction.

Assisted by: Codex

### CODE REVIEW

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 1]
```diff
@@ -1765,6 +1778,160 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+  AffineMap channel =
+      AffineMap::get(rank, 0, {b.getAffineDimExpr(axis)}, b.getContext());
+  SmallVector<AffineMap> maps = {identity, channel};
+  if (hasZeroPoints)
+    maps.push_back(channel);
+  maps.push_back(identity);
+  return maps;
+}
+
+static LogicalResult getPerChannelAxis(Operation *op, Value axisValue,
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: nit: Maybe a personal preference but can this be changed to `FailureOr<int64_t>` for readability?

Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: nit: Maybe a personal preference but can this be changed to `FailureOr<int64_t>` for readability?

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 2]
```diff
@@ -376,6 +376,68 @@ Value createRemainderPayload(OpBuilder &b, Location loc,
...
+    else
+      input = arith::ExtSIOp::create(b, loc, computeIntType, input);
+  }
+  if (zeroPoint) {
+    Type convertedType = converter->convertType(zeroPoint.getType());
+    if (zeroPoint.getType() != convertedType)
+      zeroPoint = converter->materializeTargetConversion(b, loc, convertedType,
+                                                         zeroPoint);
+    if (zeroPoint.getType() != computeIntType)
+      zeroPoint = convertScalarToDtype(b, loc, zeroPoint, computeIntType);
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: Nice enhancement to handle extension as well.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 3]
```diff
@@ -2469,46 +2636,20 @@ class ConvertDequantizePerChannel
...
-          if (zeropointDTy.isUnsignedInteger(8)) {
-            zeropoint =
-                arith::ExtUIOp::create(b, loc, b.getI32Type(), zeropoint);
-          } else if (zeropointDTy.isSignedInteger(8)) {
-            zeropoint =
-                arith::ExtSIOp::create(b, loc, b.getI32Type(), zeropoint);
-          } else if (zeropointDTy.isInteger(64)) {
-            zeropoint =
-                arith::TruncIOp::create(b, loc, b.getI32Type(), zeropoint);
-            op->emitWarning() << "truncated zero point from 64 to 32 bit";
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:2504]: I think this will always be hit since zero-points are `i64` and will be truncated. The new path with `convertScalarToDtype` doesn't produce this warning. I don't see value in emitting this diagnostic but wanted to mention here to understand whether the removal was intentional?

File: test/python/fx_importer/quantized_decomposed_import_test.py [hunk 4]
```diff
@@ -95,3 +95,45 @@ def forward(self, lhs_q, rhs_q):
...
+# CHECK: torch.quantized_decomposed.dequantize_per_channel
+# CHECK-NOT: torch.operator
+def test_dequantize_per_channel():
+    class DequantizePerChannel(nn.Module):
+        def forward(self, x, scales, zero_points):
+            return torch.ops.quantized_decomposed.dequantize_per_channel.default(
+                x, scales, zero_points, 0, -128, 127, torch.int8
+            )
+
+    m = fx.export_and_import(
```
Comment (sahas3) [inline test/python/fx_importer/quantized_decomposed_import_test.py:132]: Can you add e2e tests for numerical verification?

**Review by sahas3 — APPROVED** [review]
LG, the major feedback is to add e2e tests. Rest are nits/clarifications. Thanks!

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 3 above)
Comment (ziereis) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:2504]: hmm, i did not think much of it. I checked the torch decomposition https://github.com/pytorch/pytorch/blob/a57db29aa6d98118fdd27a7d509ab3525a750669/torch/ao/quantization/fx/_decomposed.py#L679 

And there the zeropoint would never get truncated and instead the input is always extend to the zeropoint bitwidth. Not sure if it would actually matter but i think we should do the same here. What do you think?

**Comment by ziereis** [comment]
Hi @sahas3 Thanks for the review, your comments should be addressed now. 

Regarding the zeropoint truncation it got me thinking that we probably should not be truncating the zeropoint at all to match the the torch semantics more closely. I added this change in the latest revision. What do you think?

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 5]
```diff
@@ -376,6 +376,73 @@ Value createRemainderPayload(OpBuilder &b, Location loc,
...
+    else
+      input = arith::ExtSIOp::create(b, loc, subtractionType, input);
+  }
+  if (zeroPoint)
+    input = arith::SubIOp::create(b, loc, input, zeroPoint);
+  Type convertedScaleType = converter->convertType(scale.getType());
+  if (scale.getType() != convertedScaleType)
+    scale = converter->materializeTargetConversion(b, loc, convertedScaleType,
+                                                   scale);
+  Value value = arith::SIToFPOp::create(b, loc, scale.getType(), input);
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp]: I think this needs a guard for `inputIsUnsigned` and use `UIToFP` in that case?

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp [hunk 6]
```diff
@@ -1765,6 +1783,159 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
...
+  maps.push_back(identity);
+  return maps;
+}
+
+static FailureOr<int64_t>
+getPerChannelAxis(Operation *op, Value axisValue, int64_t rank,
+                  ConversionPatternRewriter &rewriter) {
+  int64_t axis;
+  if (!matchPattern(axisValue, m_TorchConstantInt(&axis))) {
+    return failure();
```
Comment (sahas3) [inline lib/Conversion/TorchToLinalg/Uncategorized.cpp:1805]: `rewriter` is passed as an input but never used. Consider replacing the bare `failure` with `rewriter.notifyMatchFailure("msg")` to throw some meaningful info or remove the `rewriter` input.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 3 above)
Comment (sahas3) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp:2504]: Yes, makes sense to be consistent with PT.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/quantized_models.py [hunk 7]
```diff
@@ -281,3 +281,59 @@ def forward(self, x):
 @register_test_case(module_factory=lambda: QuantizedDecomposedQuantizePerTensor())
 def QuantizedDecomposedQuantizePerTensor_basic(module, tu: TestUtils):
     module.forward(tu.rand(4, 8))
+
+
+# ==============================================================================
+
+
+class QuantizedDecomposedDequantizePerChannel(torch.nn.Module):
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/quantized_models.py:289]: Looks like you need to add these to the ONNX xfail set to make CI green.

File: test/Conversion/TorchToLinalg/quantized-elementwise.mlir [hunk 8]
```diff
@@ -13,16 +13,15 @@
...
 // Operand order matters here: (input - zp), not (zp - input).
-// CHECK:         %[[SUB:.*]] = arith.subi %[[EXT]], %[[ZP_I32]] : i32
-// CHECK:         %[[SUBF:.*]] = arith.sitofp %[[SUB]] : i32 to f32
+// CHECK:         %[[SUB:.*]] = arith.subi %[[EXT]], %[[ZP_I64]] : i64
 // CHECK:         %[[SCALE_F64:.*]] = torch_c.to_f64 %[[SCALE_F]]
-// CHECK:         %[[SCALE_F32:.*]] = arith.truncf %[[SCALE_F64]] : f64 to f32
-// CHECK:         %[[MUL:.*]] = arith.mulf %[[SUBF]], %[[SCALE_F32]] : f32
-// CHECK:         linalg.yield %[[MUL]] : f32
+// CHECK:         %[[SUBF:.*]] = arith.sitofp %[[SUB]] : i64 to f64
+// CHECK:         %[[MUL:.*]] = arith.mulf %[[SUBF]], %[[SCALE_F64]] : f64
```
Comment (sahas3) [inline test/Conversion/TorchToLinalg/quantized-elementwise.mlir]: I think as per PT semantics, this `mulf` should happen in `f32` since the op is
`(input - zero_points) * scales` with `fp32` scales ?

**Comment by sahas3** [comment]
> Regarding the zeropoint truncation it got me thinking that we probably should not be truncating the zeropoint at all to match the the torch semantics more closely.

Yes, that makes sense to me. Thanks for making the change. I also noticed that for dequant we should be truncating back to `scale` type before multiplying to match with torch behavior. Other than that the change looks good.

**Comment by ziereis** [comment]
Hi @sahas3 Thanks for the review again. I made the change as discussed. One thing, i had some test failure when only accepting input programs where the zeropoint_width > input_width since there is some testcase that uses for a example a i8 zp. For this i made the extension that we extend both sides to the next higher multiple of 2 bit-width.

File: lib/Conversion/TorchToLinalg/Uncategorized.cpp (hunk 5 above)
Comment (ziereis) [reply lib/Conversion/TorchToLinalg/Uncategorized.cpp]: good catch, i think `inputIsUnsigned` but i think even just checking `inputIsUnsigned` wouldn't be sufficient here since input at this point can either refer to `input - zeroPoint` which would be better interpreted signed at this point even if the input was unsigned. 

However i added a check that does the `UIToFP` in the case where the `inputIsUnsigned` and there is no zeropoint

**Review by sahas3 — APPROVED** [review]
LG, thanks for the recent fixes.
